File: Habits/Helpers/DayKeys.m
//
//  DayKeys.m
//  Habits
//
//  Created by Michael Forrest on 16/07/2014.
//  Copyright (c) 2014 Good To Hear. All rights reserved.
//

#import "DayKeys.h"
#import "TimeHelper.h"
NSDateFormatter * dateKeyFormatter(){
    static NSDateFormatter * formatter = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        formatter = [NSDateFormatter new];
        [formatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0]];
        formatter.dateFormat = @"yyyy-MM-dd";
    });
    return formatter;
}
NSString * dayKey(NSDate* date){
    NSDateFormatter * formatter = dateKeyFormatter();
    NSDateComponents * components = [[NSCalendar currentCalendar] components:NSYearCalendarUnit|NSMonthCalendarUnit|NSDayCalendarUnit fromDate:date];
    static NSTimeZone * gmt = nil;
    static NSCalendar * calendar = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        gmt = [NSTimeZone timeZoneForSecondsFromGMT:0];
        calendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSGregorianCalendar];
        calendar.timeZone = gmt;
    });
    date = [calendar dateFromComponents:components];
    return [formatter stringFromDate:date];
}
NSDate * dateFromKey(NSString * key){
    static NSCalendar * calendar = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSTimeZone * gmt = [NSTimeZone timeZoneForSecondsFromGMT:0];
        calendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSGregorianCalendar];
        calendar.timeZone = gmt;
    });
    NSDateComponents * components = [NSDateComponents new];
    NSArray * d = [key componentsSeparatedByString:@"-"];
    components.year = [d[0] integerValue];
    components.month = [d[1] integerValue];
    components.day = [d[2] integerValue];
    NSDate * result = [calendar dateFromComponents:components];
    return result;

}
@implementation DayKeys

+(NSDate *)dateFromKey:(NSString *)key {
    return dateFromKey(key);
}
+(NSString *)keyFromDate:(NSDate *)date{
    return dayKey(date);
}
@end

File: HabitsCommon/DayKeys.swift
//
//  DayKeys.swift
//  Habits
//
//  Created by Michael Forrest on 28/11/2015.
//  Copyright © 2015 Good To Hear. All rights reserved.
//

import UIKit

private func createDateKeyFormatter()->DateFormatter{
    let formatter = DateFormatter()
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter
}
private let dateKeyFormatter = createDateKeyFormatter()
private let gmt = TimeZone(secondsFromGMT: 0)
private let gmtCalendar = Calendar(identifier: Calendar.current.identifier)

func dayKey(_ date:Date)->String{
    let formatter = dateKeyFormatter
    let components = (Calendar.current as NSCalendar).components([.year, .month, .day], from: date)
    var calendar = gmtCalendar// else { return "" }
    calendar.timeZone = gmt!
    if let date = calendar.date(from: components){
        return formatter.string(from: date)
    }else{
        return ""
    }
}
func dateFromKey(_ key:String)->Date?{
    return dateKeyFormatter.date(from: key)
}
func _dateFromKey(_ key:String)->Date?{
    return dateFromKey(key)
}
func weekdayNameOfWeekdayComponent(_ weekday:Int)->String{
    return [
        "sun", "mon", "tue", "wed", "thu", "fri", "sat"
    ][weekday - 1]
}
func weekdayOfDate(_ date:Date)->String{
    let components = (Calendar.current as NSCalendar).components(.weekday, from: date)
    return weekdayNameOfWeekdayComponent(components.weekday!)
}
@objc open class DayKeys: NSObject {
    @objc static public func convertKeyToDate(_ key:String)->Date?{
        return _dateFromKey(key)
    }
    @objc static public func convertDateToKey(_ date:Date)->String{
        return dayKey(date)
    }
}
